packages/rs-drive-abci/src/execution/platform_events/protocol_upgrade/upgrade_protocol_version/v0/mod.rs

```diff
@@ -0,0 +1,53 @@
+use crate::error::Error;
+use crate::platform_types::platform::Platform;
+use crate::platform_types::platform_state::v0::PlatformStateV0Methods;
+use crate::platform_types::platform_state::PlatformState;
+use dpp::version::PlatformVersion;
+use drive::grovedb::Transaction;
+
+impl<C> Platform<C> {
+    /// Sets current protocol version and next epoch protocol version to block platform state
+    ///
+    /// It takes five parameters:
+    /// * `epoch_info`: Information about the current epoch.
+    /// * `last_committed_platform_state`: The last committed state of the platform.
+    /// * `block_platform_state`: The current state of the platform.
+    /// * `transaction`: The current transaction.
+    /// * `platform_version`: The current version of the platform.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the previous block protocol version does not match the current block protocol version not on epoch change
+    pub(super) fn upgrade_protocol_version_v0(
+        &self,
+        last_committed_platform_state: &PlatformState,
+        block_platform_state: &mut PlatformState,
+        transaction: &Transaction,
+        platform_version: &PlatformVersion,
+    ) -> Result<(), Error> {
+        // Determine a new protocol version for the next epoch if enough proposers voted
+        // otherwise keep the current one
+
+        let hpmn_list_len = last_committed_platform_state.hpmn_list_len() as u32;
+
+        let next_epoch_protocol_version =
+            self.check_for_desired_protocol_upgrade(hpmn_list_len, platform_version)?;
+
+        if let Some(protocol_version) = next_epoch_protocol_version {
+            block_platform_state.set_next_epoch_protocol_version(protocol_version);
+        }
+
+        // Since we are starting a new epoch we need to drop previously
+        // proposed versions
+
+        let mut protocol_versions_counter = self.drive.cache.protocol_versions_counter.write();
```

Comment:
What will happen if we have multiple rounds? We are also querying from the global cache.

Reply:
Fixed